pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,380 @@
+	switch r.(type) {
+	case *os.File:
+		f, _ := r.(*os.File)
+		if fileInfo, err := f.Stat(); err == nil {
+			s := fileInfo.Size()
+			return int(math.Floor(float64(s) / alioss.MaxPartSize)), s % alioss.MaxPartSize, nil
+		}
+	case *strings.Reader:
+		f, _ := r.(*strings.Reader)
+		return int(math.Floor(float64(f.Size()) / alioss.MaxPartSize)), f.Size() % alioss.MaxPartSize, nil
+	}
+	return -1, 0, errors.New("unsupported implement of io.Reader")
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	chunksnum, lastslice, err := calculateChunks(name, r)
+	if err != nil {
+		return err
+	}
+
+	ncloser := ioutil.NopCloser(r)
+	switch chunksnum {
+	case 0:
+		if err := b.bucket.PutObject(name, ncloser); err != nil {
+			return errors.Wrap(err, "failed to upload oss object")
+		}
+	default:
+		{
+			init, err := b.bucket.InitiateMultipartUpload(name)
+			if err != nil {
+				return errors.Wrap(err, "failed to initiate multi-part upload")
+			}
+			chunk := 0
+			uploadEveryPart := func(everypartsize int64, cnk int) (alioss.UploadPart, error) {
+				prt, err := b.bucket.UploadPart(init, ncloser, everypartsize, cnk)
+				if err != nil {
+					if err := b.bucket.AbortMultipartUpload(init); err != nil {
+						return prt, errors.Wrap(err, "failed to upload multi-part chunk")
+					}
+				}
+				return prt, nil
+			}
+			var parts []alioss.UploadPart
+			for ; chunk < chunksnum; chunk++ {
+				part, err := uploadEveryPart(alioss.MaxPartSize, chunk+1)
+				if err != nil {
+					return err
+				}
+				parts = append(parts, part)
+			}
+			if lastslice != 0 {
+				part, err := uploadEveryPart(lastslice, chunksnum+1)
+				if err != nil {
+					return errors.Wrap(err, "failed to upload the last chunk")
+				}
+				parts = append(parts, part)
+			}
+			_, err = b.bucket.CompleteMultipartUpload(init, parts)
+			if err != nil {
+				return errors.Wrap(err, "failed to set multi-part upload completive")
+			}
+		}
+	}
+	return nil
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if err := b.bucket.DeleteObject(name); err != nil {
+		return errors.Wrap(err, "delete oss object")
+	}
+	return nil
+}
+
+func configFromEnv() Config {
+	c := Config{
+		Endpoint:        os.Getenv("ALIYUNOSS_ENDPOINT"),
+		Bucket:          os.Getenv("ALIYUNOSS_BUCKET"),
+		AccessKeyID:     os.Getenv("ALIYUNOSS_ACCESS_KEY_ID"),
+		AccessKeySecret: os.Getenv("ALIYUNOSS_ACCESS_KEY_SECRET"),
+	}
+	return c
+}
+
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	config, err := func(conf []byte) (Config, error) {
+		var config Config
+		if err := yaml.Unmarshal(conf, &config); err != nil {
+			return Config{}, err
+		}
+
+		return config, nil
+	}(conf)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "parse aliyun oss config file failed")
+	}
```

Comment:
Let's get rid of function creation and simplify this to:
```
	var config Config
		if err := yaml.Unmarshal(conf, &config); err != nil {
			return Config{}, errors.Wrap(err, "parse aliyun oss config file failed")
		}
```

Reply:
OK